src/S3/S3Transfer/AbstractMultipartUploader.php

```diff
@@ -0,0 +1,493 @@
+    public const PART_MAX_SIZE = 5 * 1024 * 1024 * 1024; // 5 GiB
+    public const PART_MAX_NUM = 10000;
+    public const DEFAULT_CHECKSUM_CALCULATION_ALGORITHM = 'crc32';
+    private const CHECKSUM_TYPE_FULL_OBJECT = 'FULL_OBJECT';
+
+    /** @var S3ClientInterface */
+    protected readonly S3ClientInterface $s3Client;
+
+    /** @var array */
+    protected readonly array $requestArgs;
+
+    /** @var array */
+    protected readonly array $config;
+
+    /** @var string|null */
+    protected string|null $uploadId;
+
+    /** @var array */
+    protected array $parts;
+
+    /** @var array */
+    protected array $onCompletionCallbacks = [];
+
+    /** @var TransferListenerNotifier|null */
+    protected ?TransferListenerNotifier $listenerNotifier;
+
+    /** Tracking Members */
+    /** @var TransferProgressSnapshot|null */
+    protected ?TransferProgressSnapshot $currentSnapshot;
+
+    /**
+     * For custom or default checksum.
+     *
+     * @var string|null
+     */
+    protected ?string $requestChecksum;
+
+    /**
+     * This will be used for custom or default checksum.
+     *
+     * @var string|null
+     */
+    protected ?string $requestChecksumAlgorithm;
+
+    /** @var bool */
+    private bool $isFullObjectChecksum;
+
+    /**
+     * @param S3ClientInterface $s3Client
+     * @param array $requestArgs
+     * @param array $config
+     * - target_part_size_bytes: (int, optional)
+     * - request_checksum_calculation: (string, optional)
+     * - concurrency: (int, optional)
+     * @param string|null $uploadId
+     * @param array $parts
+     * @param TransferProgressSnapshot|null $currentSnapshot
+     * @param TransferListenerNotifier|null $listenerNotifier
+     */
+    public function __construct(
+        S3ClientInterface $s3Client,
+        array $requestArgs,
+        array $config = [],
+        ?string $uploadId = null,
+        array $parts = [],
+        ?TransferProgressSnapshot $currentSnapshot = null,
+        ?TransferListenerNotifier $listenerNotifier = null,
+    ) {
+        $this->s3Client = $s3Client;
+        $this->requestArgs = $requestArgs;
+        $this->validateConfig($config);
+        $this->config = $config;
+        $this->uploadId = $uploadId;
+        $this->parts = $parts;
+        $this->currentSnapshot = $currentSnapshot;
+        $this->listenerNotifier = $listenerNotifier;
+        $this->isFullObjectChecksum = false;
+    }
+
+    /**
+     * @param array $config
+     *
+     * @return void
+     */
+    protected function validateConfig(array &$config): void
+    {
+        if (!isset($config['target_part_size_bytes'])) {
+            $config['target_part_size_bytes'] = S3TransferManagerConfig::DEFAULT_TARGET_PART_SIZE_BYTES;
+        }
+
+        if (!isset($config['concurrency'])) {
+            $config['concurrency'] = S3TransferManagerConfig::DEFAULT_CONCURRENCY;
+        }
+
+        if (!isset($config['request_checksum_calculation'])) {
```

Comment:
this validation could take place in the `MultipartUploader` constructor since checksum calculation only happens for the upload workflow and not copy.  Same goes for any other checksum-related logic